akka-persistence/src/main/scala/akka/persistence/journal/WriteJournalBase.scala

```diff
@@ -33,11 +32,17 @@ private[akka] trait WriteJournalBase {
   private[akka] final def adaptToJournal(repr: PersistentRepr): PersistentRepr = {
     val payload = repr.payload
     val adapter = eventAdapters.get(payload.getClass)
-    val manifest = adapter.manifest(payload)
 
-    repr
-      .withPayload(adapter.toJournal(payload))
-      .withManifest(manifest)
+    val repr2 = repr.withPayload(adapter.toJournal(payload))
+
+    // IdentityEventAdapter returns "" as manifest and normally the incoming PersistentRepr
+    // doesn't have an assigned manifest, but when WriteMessages is sent directly to the
+    // journal for testing purposes we want to preserve the original manifest instead of
+    // letting IdentityEventAdapter clearing it out.
+    if (adapter == IdentityEventAdapter || adapter.isInstanceOf[NoopWriteEventAdapter])
+      repr2
+    else
+      repr2.withManifest(adapter.manifest(payload))
```

Comment:
I think it can be worth avoiding the allocation from `withPayload` when it's `IdentityEventAdapter`, i.e.

    if (adapter == IdentityEventAdapter || adapter.isInstanceOf[NoopWriteEventAdapter])
      repr
    else
      repr.withPayload(adapter.toJournal(payload)).withManifest(adapter.manifest(payload))